### ISSUE TITLE
Better handling of unexpected operations in bounds-checking

### ISSUE THREAD

**Issue by jadephilipoom** [mit-plv/fiat-crypto#349]
If `ZRange.ident.interp` returns `None` for the bounds of some operation in my code, it would be nice to get some message saying which operation it was--currently I just get "not unifiable" from `solve_rop` and I'm struggling to make it show me where exactly the `None` comes from.

**Comment by JasonGross** [comment]
This would be a nontrivial amount of work, and I'm not convinced it'll be worth it.  The way to do it is to replace the `option` with an error type, and push that through the rest of the code ... but `nth_default` is the only operation that sometimes produces `Some` and sometimes produces `None`, and, beyond that, there's a list of the operations that produce `None` here:
https://github.com/mit-plv/fiat-crypto/blob/a90c2b5eeed88e611b241dc1a52cb94f92bb40e7/src/Experiments/SimplyTypedArithmetic.v#L4552-L4573

It should only take 5 minutes to write a procedure that lets you know if a given expression uses any of those (either in Ltac or in Gallina).

(I'm not opposed to having better error messages, I am just hesitant to spend the 3-6 hours I expect it take to write the code that pushes the relevant error messages through partial reduction :-/ )

**Comment by jadephilipoom** [comment]
Hmm, okay. No worries if it's that much work. I thought it might be easier.

I'm curious about what this 5-minute procedure is, though. I figured out where my problem was coming from[0] by printing the nested ltacs in `solve_rop`, painstakingly applying them until I had the reified term in view, making a simpler example with the same problem so the term wasn't huge, and then computing it step by step until I could actually see all the operations. It took me about 3 hours. I'm describing the whole process because I'm wondering if you have some more general advice about where to start with debugging when I have opaque problems with the pipeline (e.g. "there's a None in my bounds", "cache_reify takes ages", or "the bounds aren't tight enough").

[0] I think `Z.cast2` should have a case in `partial.bounds.ident.extract'` [here](https://github.com/mit-plv/fiat-crypto/blob/master/src/Experiments/SimplyTypedArithmetic.v#L5584). Is there a reason it currently doesn't?

**Comment by JasonGross** [comment]
The 5-minute procedure only works if you have the reduced reified term in hand, and it wouldn't've caught [0]

But, uh, I could make the error messages for bad bounds include the reduced syntax tree, post-bounds-analysis, if that'd be useful?

Re [0]: It's because I was operating under the (now clearly mistaken) assumption that the relevant output types were `Z` and `list Z`, and not `Z * Z`.  Feel free to add
```coq
| ident.Z_cast2 range => fun _ => ZRange.type.option.Some range
```
or
```coq
| ident.Z_cast2 (r1, r2) => fun _ => (Some r1, Some r2)
```
(they are the same)

**Comment by JasonGross** [comment]
Oh, urg, getting "not unifiable" is also not supposed to happen.  You're supposed to get a unification failure that contains a description of the error in the term that's not unifying.  Let me fix that.

**Comment by jadephilipoom** [comment]
Closed by #351
